File: day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func part1() {
	f, err := os.Open("d5.in")
	if err != nil {
		panic(err)
	}

	var (
		ins []int
		sc  = bufio.NewScanner(f)
	)
	for sc.Scan() {
		offset, err := strconv.Atoi(sc.Text())
		if err != nil {
			panic(err)
		}
		ins = append(ins, offset)
	}

	var jumps, i int
	for i != len(ins) {
		off := ins[i]
		ins[i]++
		i += off
		jumps++
	}
	fmt.Println(jumps)
}

func part2() {
	f, err := os.Open("d5.in")
	if err != nil {
		panic(err)
	}

	var (
		ins []int
		sc  = bufio.NewScanner(f)
	)
	for sc.Scan() {
		offset, err := strconv.Atoi(sc.Text())
		if err != nil {
			panic(err)
		}
		ins = append(ins, offset)
	}

	var jumps, i int
	for i != len(ins) {
		off := ins[i]
		if off >= 3 {
			ins[i]--
		} else {
			ins[i]++
		}
		i += off
		jumps++
	}
	fmt.Println(jumps)
}

func main() {
	part1()
	part2()
}

File: day5/day5.py
def part1():
    with open("d5.in") as f:
        ins = [int(l.strip()) for l in f]
        ins.append(None)

        i = 0
        jumps = 0
        while ins[i] is not None:
            ins[i], i = ins[i]+1, i+ins[i]
            jumps += 1
        print(jumps)

def part2():
    with open("d5.in") as f:
        ins = [int(l.strip()) for l in f]
        ins.append(None)

        i = 0
        jumps = 0
        while ins[i] is not None:
            ins[i], i = ins[i]-1 if ins[i] >= 3 else ins[i]+1, i+ins[i]
            jumps += 1
        print(jumps)


if __name__ == "__main__":
    part1()
    part2()